network/simple-onchain-discovery/src/lib.rs

```diff
@@ -45,95 +50,136 @@ pub static EVENT_PROCESSING_LOOP_BUSY_DURATION_S: Lazy<DurationHistogram> = Lazy
-        RoleType::Validator => config.validator_network_identity_public_key,
-        RoleType::FullNode => config.full_node_network_identity_public_key,
-    }
+fn decrypt_validator_netaddr(
+    root_key_map: &HashMap<RootKeyVersion, RootKey>,
+    account: &AccountAddress,
+    addr_idx: u32,
+    raw_enc_addr: RawEncNetworkAddress,
+) -> Result<RawNetworkAddress> {
+    let enc_addr = EncNetworkAddress::try_from(&raw_enc_addr)
+        .map_err(anyhow::Error::new)
+        .with_context(|| {
+            format_err!(
+                "error deserializing encrypted network address: {:?}",
+                raw_enc_addr
+            )
+        })?;
+
+    let key_version = enc_addr.root_key_version();
+    let root_key = root_key_map
+        .get(&key_version)
+        .ok_or_else(|| format_err!("no root_key for version: {}", key_version))?;
+    let raw_addr = enc_addr.decrypt(root_key, account, addr_idx)?;
+    Ok(raw_addr)
 }
 
 /// Extracts a set of ConnectivityRequests from a ValidatorSet which are appropriate for a network with type role.
-fn extract_updates(role: RoleType, node_set: ValidatorSet) -> Vec<ConnectivityRequest> {
-    let node_list = node_set.payload().to_vec();
-
-    let mut updates = Vec::new();
+fn extract_updates(
+    role: RoleType,
+    root_key_map: &HashMap<RootKeyVersion, RootKey>,
+    node_set: ValidatorSet,
+) -> Vec<ConnectivityRequest> {
+    // TODO(philiphayes): remove this after removing explicit pubkey field
+    let explicit_pubkeys: HashMap<_, _> = node_set
+        .payload()
+        .iter()
+        .map(|info| {
+            let peer_id = *info.account_address();
+            let pubkey = match role {
+                RoleType::Validator => info.config().validator_network_identity_public_key,
+                RoleType::FullNode => info.config().full_node_network_identity_public_key,
+            };
+            (peer_id, pubkey)
+        })
+        .collect();
 
     // Collect the set of address updates.
-    let address_map = node_list
+    let new_peer_addrs: HashMap<_, _> = node_set
+        .into_iter()
+        .map(|info| {
+            let peer_id = *info.account_address();
+            let config = info.into_config();
+            // only one field currently, though this will change soon
+            let addr_idx = 0;
+
+            let raw_addr_res = match role {
+                RoleType::Validator => {
+                    let raw_enc_addr = config.validator_network_address;
+                    decrypt_validator_netaddr(&root_key_map, &peer_id, addr_idx, raw_enc_addr)
+                }
+                RoleType::FullNode => Ok(config.full_node_network_address),
+            };
+
+            let addr_res = raw_addr_res.and_then(|raw_addr| {
+                let addr = NetworkAddress::try_from(&raw_addr)
+                    .map_err(anyhow::Error::new)
+                    .with_context(|| {
+                        format_err!("error deserializing network address: {:?}", raw_addr)
```

Comment:
It seems like a lot of nesting in one map function.  Do you think it's possible / worth it to break it up?